File: AddBinary.cpp
#include <bits/stdc++.h>
using namespace std;

string addBinary(string a, string b) {
    string ans = "";
    int carry = 0;

    reverse(a.begin(), a.end());
    reverse(b.begin(), b.end());  // reverse strings

    int n = max(a.length(), b.length());

    int digit1, digit2;
    for (int i = 0; i < n; i++) {

        if(i < a.length()){
            digit1 = a[i] - '0';// char -> int
        }
        else{
            digit1 = 0;
        }

        if(i < b.length()){
            digit2 = b[i] - '0';
        }
        else{
            digit2 = 0;
        }
        int total = digit1 + digit2 + carry;

        ans = char((total % 2) + '0') + ans;  // current bit
        carry = total / 2;                   // carry
    }

    if (carry == 1) {
        ans = '1' + ans;  // leftover carry
    }

    return ans;
}

int main() {
    string a = "11";
    string b = "01";

    string res = addBinary(a, b);
    cout << res << endl;

    return 0;
}

// string addBinary(string a, string b) {
//     int i = a.length() - 1;
//     int j = b.length() - 1;
//     int carry = 0;

//     string ans = "";

//     while (i >= 0 || j >= 0 || carry) {
//         int sum = carry;

//         if (i >= 0) {
//             sum += a[i] - '0';  // convert '0'/'1' → 0/1
//             i--;
//         }

//         if (j >= 0) {
//             sum += b[j] - '0';
//             j--;
//         }

//         char ch = sum % 2 + '0';
//         ans = ch + ans;
//         carry = sum / 2;
//     }

//     return ans;
// }

File: AddBinary.py
        
def addBinary(a: str, b: str) -> str:
    ans = ""
    carry = 0

    a, b = a[::-1], b[::-1]  # reverse strings

    for i in range(max(len(a), len(b))):

        if i < len(a):
            digit1 = ord(a[i]) - ord('0') #convert char -> int
        else:
            digit1 = 0

        if i < len(b):
            digit2 = ord(b[i]) - ord('0')
        else:
            digit2 = 0

        total = digit1 + digit2 + carry

        value = str(total % 2) # 1 + 1 = 2 (2 % 2 -> value = 0)
        ans = value + ans

        carry = total // 2 # 1 + 1 = 2 (2 / 2 -> carry = 1)

    if carry == 1:
        ans = "1" + ans

    return ans


def main():
    a = "11"
    b = "01"

    res = addBinary(a, b)
    print(res)

if __name__ == "__main__":
    main()



# ------- shorter method ----------------
# def addBinary(a: str, b: str) -> str:
#     ans = ""
#     carry = 0

#     a, b = a[::-1], b[::-1]  # reverse to process from LSB

#     for i in range(max(len(a), len(b))):
#         digit1 = int(a[i]) if i < len(a) else 0
#         digit2 = int(b[i]) if i < len(b) else 0

#         total = digit1 + digit2 + carry

#         ans = str(total % 2) + ans   # current bit
#         carry = total // 2           # carry

#     if carry:
#         ans = "1" + ans              # leftover carry

#     return ans

# def main():
#     a = "11"
#     b = "01"

#     res = addBinary(a, b)
#     print(res)
    
# if __name__ == "__main__":
#     main()